storage/leia.go

```diff
@@ -0,0 +1,300 @@
+	"github.com/nuts-foundation/go-stoabs"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/storage/log"
+)
+
+type KVBackedLeiaStore interface {
+	leia.Store
+	AddConfiguration(config LeiaBackupConfiguration)
+	// HandleRestore migrates the data from the backup store to the leia.Store if needed.
+	// It's up to the caller to create the indices on the leia.Collections first before calling this method.
+	HandleRestore() error
+}
+
+type kvBackedLeiaStore struct {
+	store               leia.Store
+	backup              stoabs.KVStore
+	collectionConfigSet map[string]LeiaBackupConfiguration
+}
+
+// NewKVBackedLeiaStore creates a wrapper around a leia.Store that uses a stoabs.KVStore as backup for any documents added.
+// The backup store is not closed when Close is called. The leia.Store is closed when Close is called.
+func NewKVBackedLeiaStore(store leia.Store, backup stoabs.KVStore) (KVBackedLeiaStore, error) {
+	return &kvBackedLeiaStore{
+		store:               store,
+		backup:              backup,
+		collectionConfigSet: map[string]LeiaBackupConfiguration{},
+	}, nil
+}
+
+type CollectionType int
+
+const (
+	JSONLDCollectionType CollectionType = iota
+	JSONCollectionType   CollectionType = iota
+)
+
+type LeiaBackupConfiguration struct {
+	// CollectionName is the name of the collection in the leia.Store.
+	CollectionName string
+	CollectionType CollectionType
+	// BackupShelf is the name of the shelf in the backup store.
+	BackupShelf string
+	// SearchQuery is used to fill the backup shelf if not present.
+	SearchQuery leia.QueryPath
+}
+
+type kvBackedCollection struct {
+	backup         stoabs.KVStore
+	config         LeiaBackupConfiguration
+	collectionType CollectionType
+	underlying     leia.Collection
+}
+
+func (k *kvBackedLeiaStore) AddConfiguration(config LeiaBackupConfiguration) {
+	k.collectionConfigSet[config.CollectionName] = config
+}
+
+func (k *kvBackedLeiaStore) HandleRestore() error {
+	// leia indices have been added, so the collection names have been added to the collectionConfigSet.
+	// Loop over this set to check if the backup store contains any documents for these collections and add them to the leia.Store if the named collection is empty there.
+	for _, config := range k.collectionConfigSet {
+		if err := k.handleRestore(config); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (k *kvBackedLeiaStore) handleRestore(config LeiaBackupConfiguration) error {
+	backupPresent := k.backupStorePresent(config.BackupShelf)
+	var collection leia.Collection
+	switch config.CollectionType {
+	case JSONLDCollectionType:
+		collection = k.store.JSONLDCollection(config.CollectionName)
+	case JSONCollectionType:
+		collection = k.store.JSONCollection(config.CollectionName)
+	default:
+		return errors.New("unknown collection type")
+	}
+	storePresent := storePresent(collection, config)
+
+	if backupPresent && storePresent {
+		// both are filled => normal operation, done
+		return nil
```

Comment:
Although this didn't change, could it not be the store is not populated completely, and the restore is still required?

Reply:
There's no way to detect that other then the node didn't start on the previous run due to an error.